File: modeler/src/main/java/net/sayaya/document/modeler/model/ModelRouter.java
package net.sayaya.document.modeler.model;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.codec.ServerSentEvent;
import org.springframework.web.reactive.function.BodyInserters;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Mono;

import static org.springframework.web.reactive.function.server.RequestPredicates.DELETE;
import static org.springframework.web.reactive.function.server.RequestPredicates.GET;
import static org.springframework.web.reactive.function.server.RequestPredicates.PUT;
import static org.springframework.web.reactive.function.server.RouterFunctions.route;

@Configuration
public class ModelRouter {
	private final ModelHandler handler;
	public ModelRouter(ModelHandler handler) {
		this.handler = handler;
	}
	@Bean
	public RouterFunction<ServerResponse> modelRouterInstance() {
		return route(GET("/models"), this::findModels)
				.andRoute(PUT("/models/{id}"), this::createModels)
				.andRoute(DELETE("/models/{id}"), this::removeModels)
				.andRoute(GET("/models/changes"), this::subscribeModel);
	}
	private Mono<ServerResponse> findModels(ServerRequest request) {
		return handler.list().collectList()
				.flatMap(list->ServerResponse.ok().contentType(MediaType.APPLICATION_JSON).bodyValue(list))
				.switchIfEmpty(ServerResponse.noContent().build())
				.onErrorResume(e->ServerResponse.badRequest().bodyValue(e.getMessage()));
	}
	private Mono<ServerResponse> createModels(ServerRequest request) {
		String id = request.pathVariable("id");
		return handler.create(id)
				.flatMap(obj->ServerResponse.ok().build())
				.switchIfEmpty(ServerResponse.status(HttpStatus.CONFLICT).build())
				.onErrorResume(e->ServerResponse.badRequest().bodyValue(e.getMessage()));
	}
	private Mono<ServerResponse> removeModels(ServerRequest request) {
		String id = request.pathVariable("id");
		return handler.remove(id)
				.flatMap(obj->ServerResponse.ok().build())
				.onErrorResume(e->{
					e.printStackTrace();
					return ServerResponse.badRequest().bodyValue(e.getMessage());
				});
	}
	private Mono<ServerResponse> subscribeModel(ServerRequest request) {
		return ServerResponse.ok().contentType(MediaType.TEXT_EVENT_STREAM)
				.body(BodyInserters.fromServerSentEvents(handler.subscribe()
						.map(msg->ServerSentEvent.builder(msg.data())
								.event(msg.type().name())
								.id(msg.data().name())
								.build())));
	}
}

File: modeler/src/main/kotlin/net/sayaya/document/modeler/model/ModelRouter.kt
package net.sayaya.document.modeler.model

import net.sayaya.document.data.MessageModel
import net.sayaya.document.data.Model
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.codec.ServerSentEvent
import org.springframework.web.reactive.function.BodyInserters
import org.springframework.web.reactive.function.server.*
import reactor.core.publisher.Mono

@Configuration
class ModelRouter(private val handler: ModelHandler) {
    @Bean
    fun modelRouterInstance(): RouterFunction<ServerResponse?> {
        return RouterFunctions
            .route(RequestPredicates.GET("/models"), this::findModels)
            .andRoute(RequestPredicates.PUT("/models/{id}"), this::createModels)
            .andRoute(RequestPredicates.DELETE("/models/{id}"), this::removeModels)
            .andRoute(RequestPredicates.GET("/models/changes"), this::subscribeModel)
    }
    private fun findModels(request: ServerRequest): Mono<ServerResponse> {
        return handler.list().collectList()
            .flatMap(ServerResponse.ok().contentType(MediaType.APPLICATION_JSON)::bodyValue)
            .switchIfEmpty(ServerResponse.noContent().build())
            .onErrorResume (ServerResponse.badRequest()::bodyValue)
    }
    private fun createModels(request: ServerRequest): Mono<ServerResponse> {
        return handler.create(request.pathVariable("id"))
            .flatMap { ServerResponse.ok().build() }
            .switchIfEmpty(ServerResponse.status(HttpStatus.CONFLICT).build())
            .onErrorResume (ServerResponse.badRequest()::bodyValue)
    }
    private fun removeModels(request: ServerRequest): Mono<ServerResponse> {
        return handler.remove(request.pathVariable("id"))
            .flatMap { ServerResponse.ok().build() }
            .onErrorResume { e ->
                e.printStackTrace()
                ServerResponse.badRequest().bodyValue(e.message!!)
            }
    }
    private fun subscribeModel(request: ServerRequest): Mono<ServerResponse> {
        return ServerResponse.ok().contentType(MediaType.TEXT_EVENT_STREAM)
            .body(BodyInserters.fromServerSentEvents(handler.subscribe().map {
                    msg -> ServerSentEvent.builder<Model>(msg.data).event(msg.type.name).id(msg.data.name()).build()
            }))
    }
}